staker-rewards/lib/sla-calculator.js

```diff
@@ -0,0 +1,148 @@
+import { KeepStatus, KeepTerminationCause } from "./keep.js";
+
+export default class SLACalculator {
+    constructor(
+        openedKeeps,
+        keygenFailKeeps,
+        deactivatedKeeps,
+        signatureFailKeeps
+    ) {
+        this.openedKeeps = openedKeeps
+        this.keygenFailKeeps = keygenFailKeeps
+        this.deactivatedKeeps = deactivatedKeeps
+        this.signatureFailKeeps = signatureFailKeeps
+    }
+
+    static initialize(cache, interval) {
+        // Step 1 of keygen SLA: get keeps opened within the given interval
+        const openedKeeps = cache
+            .getKeeps() // get keeps with all statuses
+            .filter(keep =>
+                // get keeps whose creation timestamps are within
+                // the given interval
+                interval.start <= keep.creationTimestamp &&
+                keep.creationTimestamp < interval.end
+            )
+
+        // Step 2 of keygen SLA: from keeps opened within the given interval,
+        // get the ones which have been eventually terminated due to keygen fail
+        const keygenFailKeeps = openedKeeps
+            .filter(keep =>
+                keep.status.name === KeepStatus.TERMINATED &&
+                keep.status.cause === KeepTerminationCause.KEYGEN_FAIL
+            )
+
+        // Step 1 of signature SLA: get keeps closed within the given interval
+        const closedKeeps = cache
+            .getKeeps(KeepStatus.CLOSED) // get closed keeps
+            .filter(keep => {
+                // get keeps whose statuses have been changed within the
+                // given interval
+                return interval.start <= keep.status.timestamp &&
+                    keep.status.timestamp < interval.end
+            })
+
+        // Step 2 of signature SLA: get keeps terminated within the
+        // given interval
+        const terminatedKeeps = cache
+            .getKeeps(KeepStatus.TERMINATED) // get terminated keeps
+            .filter(keep => {
+                // get keeps whose statuses have been changed within the
+                // given interval
+                return interval.start <= keep.status.timestamp &&
+                    keep.status.timestamp < interval.end
+            })
+
+        // Step 3 of signature SLA: Concatenate keeps closed and terminated
+        // within the given interval but exclude the keeps terminated due to
+        // keygen fail as they are not relevant for the signature SLA. This way
+        // we obtain an array of keeps whose statuses have been changed from
+        // `active` to `closed`/`terminated`. Implicitly, this means a keep
```

Comment:
> This way we obtain an array of keeps whose statuses have been changed from `active` to `closed` or from `active` to `terminated` *but not as a result of a failed key generation*.

Let's maybe make it more precise ☝️ 

(*Before doing anything here, please take a look at my comment to the code below).